strutil/env.go

```diff
@@ -0,0 +1,253 @@
+	if key == "" {
+		return "", "", fmt.Errorf("environment variable name cannot be empty: %q", entry)
+	}
+	return key, value, nil
+}
+
+// ParseRawEnvironment parsers raw environment.
+//
+// This function fails if any of the provided values are not in the form of
+// key=value or if there are duplicate keys.
+func ParseRawEnvironment(raw RawEnvironment) (*Environment, error) {
+	env := &Environment{entries: make(map[string]string, len(raw))}
+	for _, entry := range raw {
+		key, value, err := parseEnvEntry(entry)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := env.entries[key]; ok {
+			return nil, fmt.Errorf("cannot overwrite earlier value of %q", key)
+		}
+		env.entries[key] = value
+	}
+	return env, nil
+}
+
+// OSEnvironment returns the environment of the calling process.
+func OSEnvironment() (*Environment, error) {
+	return ParseRawEnvironment(os.Environ())
+}
+
+// Transform programmatically replaces all keys and values.
+//
+// If multiple keys are transformed into the same key the value of the last
+// (lexicographically) key is used. If the transformed key is empty then the
+// corresponding entry is removed.
+func (env *Environment) Transform(tr func(key, value string) (string, string)) {
+	newEntries := make(map[string]string, len(env.entries))
+	keys := make([]string, 0, len(env.entries))
+	for key := range env.entries {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		newKey, newValue := tr(key, env.entries[key])
+		if newKey != "" {
+			newEntries[newKey] = newValue
+		}
+	}
+	env.entries = newEntries
+}
+
+// Get returns the value of a given environment variable.
+func (env *Environment) Get(key string) string {
+	return env.entries[key]
+}
+
+// Contains returns true if given environment variable is set.
+func (env *Environment) Contains(key string) bool {
+	_, ok := env.entries[key]
+	return ok
+}
+
+// Del removes the given environment variable.
+func (env *Environment) Del(key string) {
+	delete(env.entries, key)
+}
+
+// Set adds or replaces the given environment variable.
+func (env *Environment) Set(key, value string) {
+	if env.entries == nil {
+		env.entries = make(map[string]string)
+	}
+	env.entries[key] = value
+}
+
+// RawEnvironment returns the raw equivalent of the given environment.
+// The returned environment is sorted lexicographically by variable name.
+func (env *Environment) RawEnvironment() RawEnvironment {
+	raw := make([]string, 0, len(env.entries))
+	keys := make([]string, 0, len(env.entries))
+	for key := range env.entries {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		raw = append(raw, fmt.Sprintf("%s=%s", key, env.entries[key]))
+	}
+	return raw
+}
+
+// EnvironmentDelta describes alterations to an environment.
+//
+// Differential environment can refer to existing entries by using shell-like
+// syntax $KEY or ${KEY}. Entries inside an environment delta are ordered.
+type EnvironmentDelta struct {
```

Comment:
do we need this if Environment itself becomes an OrderedMap ?

Reply:
Yes because the key difference is that Environment contains just the values and Delta effectively has expressions that refer to values. This is needed to correctly compose changes from snap overrides and combine them with actual system environment.

Obviously one could re-imagine this problem as another set of types entirely, I'm just clarifying why they are distinct.

Reply:
this can be:

type ExpandableEnv *OrderedMap

Reply:
Ended up as an embedded pointer in a struct to get the method set.